### PR TITLE
Fixing isEquivalent on Windows

### PR DESCRIPTION
**Pull request by herebcs**
On Windows isEquivalent sometimes returns true for unrelated files, which makes the scraper overwrite games already scraped. As FAT/NTFS do not have the inode concept, files having the same size and created at the same time (e.g. files having rom ids imported from fs-uae's database by a script) have the same stat64 attributes as well. Using GetFileInformationByHandle instead on Windows.

### CODE REVIEW

**Comment by pjft** [comment]
I can't test this on Windows, but thanks for putting this together. Happy holidays.

**Comment by cmitu** [comment]
I can test this on Windows - do you have a test case to try ?

**Comment by herebcs** [comment]
> I can test this on Windows - do you have a test case to try ?

I can create one. Is there a specific format you expect or a repository for test cases?

**Comment by cmitu** [comment]
Not in code, just describe how I can reproduce the problem - It's not clear from the issue opened how can I create 2 ROMs that would trigger this problem.

**Comment by herebcs** [comment]
Ok, I try to describe what to do.

- First, you need to create a few files (dummy ROMs) with any extension that one of your existing systems in es_systems.cfg recognizes. 

- The content of these files does not matter, but their size must match (e.g. 8 bytes each with random characters) and their creation data/time as well. You can set the creation date by many tools, I'm not sure if there is a terminal command on windows to do this. In real scenarios the files are created by a script roughly at the same time, that is why many of the matching dates.

- You also need to name these files in a way that the scraper can recommend a game based on the filename.

(If you are wondering why such a strange setup is relevant at all, please read [https://retropie.org.uk/forum/topic/18409/guide-getting-emulationstation-to-work-with-fs-uae](url))

- Now you need to run the scraper.
- Once the scraper found a match for all the ROMs you created just exit EmulationStation and restart it.
- If the bug is present then some of your games appear without the information the scraper added in the previous step. If you only have the dummy files for that particular system, then only the game last scraped will survive the restart.

Please let me know, if you need more information.

**Comment by cmitu** [comment]
@herebcs OK, seems simple enough. I wasn't sure about the 3rd condition (name based search). I know that some systems/emulators will always have a fixed ROM size. I'll give it a whirl, maybe it can be merged as a part of #518.

**Comment by cmitu** [comment]
@herebcs Looks fine. 
You don't have to scrape to get the error, just need to modify the metadata on the ROM entry to let ES know it has changed. 
So the problem is indeed `st_ino` being zero on Win32, hence if the other attributes are the same, the modified ROM entry is ignored and doesn't get added to the `gamelist.xml` (but it will show up again on next restart, but without any of the metadata from the previous session)>

**Comment by herebcs** [comment]
Great, thanks for testing it!

File: es-core/src/utils/FileSystemUtil.cpp [hunk 1]
```diff
@@ -656,15 +656,51 @@ namespace Utils
...
+
+						struct __stat64 info1;
+						struct __stat64 info2;
+
+						// check if stat succeeded
+						if((stat64(path1.c_str(), &info1) != 0) || (stat64(path2.c_str(), &info2) != 0))
+							return false;
+
+						// check if attributes are identical
+						return ((info1.st_dev == info2.st_dev) && (info1.st_ino == info2.st_ino) && (info1.st_size == info2.st_size) && (info1.st_mtime == info2.st_mtime));
```
Comment (joolswills) [inline es-core/src/utils/FileSystemUtil.cpp]: please can you reduce the indent of the above. 1 tab should be enough.

Comment (herebcs) [reply es-core/src/utils/FileSystemUtil.cpp]: Unfortunately I deleted the original repository that this pull request gets the source from. I don't know how to make additional changes without creating a new pull request, but that would erase the conversation that we are having here... Any idea?

Comment (cmitu) [reply es-core/src/utils/FileSystemUtil.cpp]: I think you can change the base branch of your PR - https://help.github.com/articles/changing-the-base-branch-of-a-pull-request/ - to keep this in one PR.

Comment (herebcs) [reply es-core/src/utils/FileSystemUtil.cpp]: Thanks. I tried that but unfortunately you can only change the base branch not the one which the changes are coming from.

**Comment by tomaz82** [comment]
Been trying to understand why the isEquivalent check is even needed,

`				if(nodePath == gamePath || (Utils::FileSystem::exists(nodePath) &&
				                            Utils::FileSystem::exists(gamePath) &&
				                            Utils::FileSystem::isEquivalent(nodePath, gamePath)))`

In which case would nodePath and gamePath not match, but a fixed isEquivalent return true?

**Comment by herebcs** [comment]
I believe that it tries to deal with the concept of aliases that most file systems have. Two different paths could point to the same physical file.

**Comment by tomaz82** [comment]
That was my assumption as well, so this should really be fixed by making both paths canonical then, and eliminate the need for isEquivalent, I'll write up a PR for that later today.

**Comment by hhromic** [comment]
This seems to be now fixed by #549 (by @tomaz82), perhaps can be closed then?

**Pull request closed without merging**